moon-store-mac-os/moon-store-mac-os/Domain/Models/RoleModel.swift

```diff
@@ -0,0 +1,34 @@
+//
+//  RoleModel.swift
+//  moon-store-mac-os
+//
+//  Created by Jose Luna on 12/14/24.
+//
+
+struct RoleModel: Decodable {
+    let id: Int
+    let name: String
+    let description: String
+    let createdAt: String?
+    let updatedAt: String?
+    let deletedAt: String?
+    
+    func copyWith(name: String? = nil,
+                  description: String? = nil) -> RoleModel {
+        RoleModel(id: id,
+                  name: name ?? self.name,
+                  description: description ?? self.description,
+                  createdAt: createdAt,
+                  updatedAt: updatedAt,
+                  deletedAt: deletedAt)
+    }
+    
+    static func factory(from role: RoleSwiftDataModel) -> RoleModel {
+        RoleModel(id: role.id,
+                  name: role.name,
+                  description: role.desc,
+                  createdAt: role.createdAt,
+                  updatedAt: role.updatedAt,
+                  deletedAt: role.deletedAt)
+    }
```

Comment:
For me, this is unnecesary code, Swift has its own methods for creating and copying objects.

Flutter patterns are strangers when applying on Swift code.

Reply:
@DerianCordobaPerez **Factory** pattern is not specific to Flutter to begin with; it’s a general concept that can be applied anywhere. Its main benefit lies in converting between models without overloading a structure with Codable methods and multiple encoders. You should focus more on the utility rather than the abstraction. You don’t need a cannon (implementing another Codable encoder) to kill a mosquito (solving the problem).

Perhaps the use of the prototype pattern is debatable, given that in environments like Flutter, data is immutable, and we use copyWith (prototype) to change certain properties because mutation is not allowed by default. In Swift, this is left to the discretion of the developer.

How can you generalize that it’s not used in Swift? What basis or data do you have for saying it’s not used? A quick Google search will show people doing exactly the same thing.

[How to copy objetcs in swift?](https://stackoverflow.com/questions/28174531/how-do-you-copy-objects-in-swift#28175543)

[Swift using copyWith](https://www.hackingwithswift.com/example-code/system/how-to-copy-objects-in-swift-using-copy)

Reply:
@Rexmoon You are wrong, the copy pattern is used on Swift when you use `NSZone` and `Reflection` programming.

The second article shows the correct way, you are using directly the values.

Now, if you want to create a lot of code to implement "factory" and "copy" methods is another topic.

Remind that you are using SwiftData, persistence models, you don't need copy the persistence models in normal models.

Reply:
@DerianCordobaPerez  [Copy with in native swift](https://developer.apple.com/documentation/foundation/nscopying/1410311-copy)

Reply:
@DerianCordobaPerez  Implementation here is to strike a balance between functionality and simplicity. If using factory methods and copy patterns adds clarity to the workflow and makes the codebase easier to maintain, it's worth considering. 

For example:

Another time we need to change only the name property?

Will we need to fill all other properties?